crates/verify/src/bytecode.rs

```diff
@@ -0,0 +1,574 @@
+                        "Bytecode does not exist for contract deployed on fork at address {}",
+                        contract_address
+                    );
+                }
+            }
+            None => {
+                eyre::bail!(
+                    "Failed to get runtime code for contract deployed on fork at address {}",
+                    contract_address
+                );
+            }
+        };
+
+        let onchain_runtime_code = provider
+            .get_code_at(
+                self.address,
+                Some(BlockId::Number(BlockNumberOrTag::Number(simulation_block))),
+            )
+            .await?;
+
+        // Compare the runtime bytecode with the locally built bytecode
+        let res = try_match(
+            &fork_runtime_code.bytecode,
+            &onchain_runtime_code,
+            &constructor_args,
+            &verification_type,
+            true,
+            has_metadata,
+        )?;
+
+        if res.0 {
+            println!(
+                "{} with status {}",
+                Paint::green("Runtime code matched").bold(),
+                Paint::green(res.1.unwrap()).bold()
+            );
+        } else {
+            println!(
+                "{}",
+                Paint::red(
+                    "Runtime code did not match - This may be due to varying compiler settings"
+                )
+                .bold()
+            );
+        }
+
+        Ok(())
+    }
+
+    fn build_project(&self, config: &Config) -> Result<Bytes> {
+        let project = config.project()?;
+        let mut compiler = ProjectCompiler::new();
+
+        if let Some(skip) = &self.skip {
+            if !skip.is_empty() {
+                compiler = compiler.filter(Box::new(SkipBuildFilters::new(skip.to_owned())?));
+            }
+        }
+        let output = compiler.compile(&project)?;
+
+        let artifact = output
+            .find_contract(&self.contract)
+            .ok_or_eyre("Build Error: Contract artifact not found locally")?;
+
+        let local_bytecode = artifact
+            .get_bytecode_object()
+            .ok_or_eyre("Contract artifact does not have bytecode")?;
+
+        let local_bytecode = match local_bytecode.as_ref() {
+            BytecodeObject::Bytecode(bytes) => bytes,
+            BytecodeObject::Unlinked(_) => {
+                eyre::bail!("Unlinked bytecode is not supported for verification")
+            }
+        };
+
+        Ok(local_bytecode.to_owned())
+    }
+
+    fn build_using_cache(&self, config: &Config) -> Option<Bytes> {
+        let project = config.project().ok()?;
+        let cache = project.read_cache_file().ok()?;
+        let cached_artifacts = cache.read_artifacts::<CompactContractBytecode>().ok()?;
+
+        for (key, value) in cached_artifacts {
+            let name = self.contract.name.to_owned() + ".sol";
+            if key.ends_with(name.as_str()) {
+                if let Some(artifact) = value.into_iter().next() {
```

Comment:
We should probably make this fn version-aware to allow correct artifact selection for multi-version projects